pandas/core/arrays/datetimes.py

```diff
@@ -184,36 +229,97 @@ class DatetimeArrayMixin(dtl.DatetimeLikeArrayMixin,
     # Constructors
 
     _attributes = ["freq", "tz"]
-    _tz = None
     _freq = None
 
-    @classmethod
-    def _simple_new(cls, values, freq=None, tz=None):
-        """
-        we require the we have a dtype compat for the values
-        if we are passed a non-dtype compat, then coerce using the constructor
-        """
-        assert isinstance(values, np.ndarray), type(values)
+    def __init__(self, values, dtype=_NS_DTYPE, freq=None, copy=False):
```

Comment:
Not having `tz` in the `__init__` allows us to share a signature across DatetimeArray, TimedetlaArray, and PeriodArray. We're *close* to being able to share an implementation.

Since `tz` would be redundant with `dtype`, I found it easier to be strict and just allow dtype.

For PeriodArray, `freq` is actually redundant with `dtype`, but `freq` is already there for not having to re-infer for DTA and TDA, so it's probably best to support it.

Reply:
> Can you point to an example where it allows code sharing?

By code-sharing I meant merging `DatetimeArray.__init__ and TimedeltaArray.__init__`, but we don't currently do this.

> and keep the init more straightforward

I think that whether DatetimeArray gets the (maybe) tz from `tz=` or `dtype=` is mainly irrelevant for the complexity of the `init`. We'll still need to validate that tz against the maybe tz on the `data`, and we'll still need to normalize the tz.

Reply:
And just to re-iterate, `tz` is not a "direct" property of the DatetimeArray. Rather it's a reference to `self.dtype.tz`, which is an Optional timezone.

Reply:
I agree with @TomAugspurger here, this is an internal constructor and should just accept dtype